# Newton's method (Solutions)
In this exercise you will use Newton's method for finding roots of the scalar function $f(x)$ to within a certain level of prescision (i.e., some small number $\texttt{tol}$). Recall that with Newton's method, you make a guess for the root $x_0$, and then you draw a tangent line of $f(x)$ line at $x=x_0$. You then use the root of the tangent line as an $\it{improved}$ guess of the root, which we will call $x_1$. We then draw another tangent line, now at the point $x_1$ and keep going $n$ times until your guess $x_{n}$ satifies some sort of stopping criteria. 

Newton's method reads $$x_{k+1}=x_{k}-\frac{f(x_{k})}{f'(x_{k})},$$ which is the solution to the root of the tangent line of $f(x)$ at $x=x_k$. Note that this is best implemented using a while loop. 

**Stopping criteria**: Your stopping criteria should be something like $\texttt{abs}(f(x_n))<\texttt{tol}$ and/or $\texttt{abs}(x_n-x_{n+1})<\texttt{tol}$. In addition, it is sometimes wise to add another stopping criteria in case the algorithm $\it does~not$ converge, for example 

    k=1
    while <<stopping_criteria>> and k<100:
        <<Newton iteration>>
        k = k+1
        
this will stop the loop if it hasn't converged in 100 iterations.



## a) 

Use Newton's method to calculate the roots of the test function $f(x)=\cos(x)$, which has known roots at $x = \frac{n \pi}{2}$, for some integer $m$. 

Use a tolerance of $\texttt{tol} = 10^{-10}$, and an initial guess of $x_0 = 0.5$.

Your algorithm should converge to the root $x = \frac{\pi}{2}$. 

**Solution:**

In [1]:
from math import *
x = 0.5              # Initial guess, user-specified
tol = 1e-10          # Tolerance, user-specified
k = 0                # Variable to keep track of the number of iterations

# while stopping_criteria is not stisfied and k is less than 20 itertions
while abs(cos(x)) > tol and k < 20:
    k = k + 1 
    x = x - cos(x)/(-sin(x)) # Newton iteration 
    print(f"n = {k} -- x_n = {x} -- f(x_n) = {cos(x)} ")
    
print(f"Newton's method converged in {k} iterations")

n = 1 -- x_n = 2.330487721712452 -- f(x_n) = -0.6886977242318095 
n = 2 -- x_n = 1.3806234748302173 -- f(x_n) = 0.18902863351567445 
n = 3 -- x_n = 1.5731225635727122 -- f(x_n) = -0.0023262346797918885 
n = 4 -- x_n = 1.570796322598839 -- f(x_n) = 4.1960575626922446e-09 
n = 5 -- x_n = 1.5707963267948966 -- f(x_n) = 6.123233995736766e-17 
Newton's method converged in 5 iterations


### i) ###
For the stopping criteria $\texttt{abs}(f(x_n))<\texttt{tol}$, how many iterations does it take for Newton's method to converge to the root? 

**Solution:**

5

### ii) ### 
What happens when you use the initial guess of $x_0 = 0$? Can you explain your observation? (Note: if you have written your code correctly, something $\it should$ go wrong.)

**Solution:**

At $x = 0$, we have $f'(0) = -\sin(0) = 0$, which means that in the first Newton iteration $$x_{1} = x_0 + \frac{f(x_0)}{f'(x_0)} = 0 + \frac{\cos(0)}{-\sin(0)} = \frac
{1}{0},$$  i.e. we are dividing by zero which we can't do. This is equivalent to finding a root of a tangent line that is parallel to the horizontal axis (i.e. it will never intercept... well technically parallel lines intercept at infinity but this not relevant for numerics!)

Below is the same code as above but with our intial guess at zero $x_0 = 0$. Note that we get a "ZeroDivisionError" message. 

In [2]:
from math import *
x = 0       # Initial guess        
tol = 1e-10 # Tolerance (user-specified)
k = 0       # Variable to keep track of the number of iteration

while abs(cos(x)) > tol and k < 20:
    k = k + 1 
    x = x - cos(x)/(-sin(x)) # Newton iteration 
    print(f"n = {k} -- x_n = {x} -- f(x_n) = {cos(x)}")



ZeroDivisionError: float division by zero

### iii)

What happens when you use a tolerance of $\texttt{tol} = 10^{-18}$ and $x_0=0.5$? Does the algorithm converge? Can you explain your observation?

**Solution:**

The iterations fail to converge. Setting $\texttt{tol}=10^{-18}$ is a bad idea because this is below machine precision, which means round-off errors are greater than the accuracy that you are asking of your solution. 

This is illustrated by the below code, where we observe that the Newton iterations converge to a certain root but is unable to get lower than $f(x_n) = 6.12323\times 10^{-17}$ as round-off errors dominate at this scale of precision. 

In [3]:
from math import *
x = 0.5    # Initial guess        
tol = 1e-18 # Tolerance (user-specified)
k = 0       # Variable to keep track of the number of iteration

while abs(cos(x)) > tol and k < 20:
    k = k + 1 
    x = x - cos(x)/(-sin(x)) # Newton iteration 
    print(f"n = {k} -- x_n = {x} -- f(x_n) = {cos(x)}")

n = 1 -- x_n = 2.330487721712452 -- f(x_n) = -0.6886977242318095
n = 2 -- x_n = 1.3806234748302173 -- f(x_n) = 0.18902863351567445
n = 3 -- x_n = 1.5731225635727122 -- f(x_n) = -0.0023262346797918885
n = 4 -- x_n = 1.570796322598839 -- f(x_n) = 4.1960575626922446e-09
n = 5 -- x_n = 1.5707963267948966 -- f(x_n) = 6.123233995736766e-17
n = 6 -- x_n = 1.5707963267948966 -- f(x_n) = 6.123233995736766e-17
n = 7 -- x_n = 1.5707963267948966 -- f(x_n) = 6.123233995736766e-17
n = 8 -- x_n = 1.5707963267948966 -- f(x_n) = 6.123233995736766e-17
n = 9 -- x_n = 1.5707963267948966 -- f(x_n) = 6.123233995736766e-17
n = 10 -- x_n = 1.5707963267948966 -- f(x_n) = 6.123233995736766e-17
n = 11 -- x_n = 1.5707963267948966 -- f(x_n) = 6.123233995736766e-17
n = 12 -- x_n = 1.5707963267948966 -- f(x_n) = 6.123233995736766e-17
n = 13 -- x_n = 1.5707963267948966 -- f(x_n) = 6.123233995736766e-17
n = 14 -- x_n = 1.5707963267948966 -- f(x_n) = 6.123233995736766e-17
n = 15 -- x_n = 1.5707963267948966 -- f(x_n) = 

## b)

Now we will try and find a solution to the following function $$x{{\rm e}^{- \left( \sin \left( x/2 \right)  \right) ^{2}}}=3/2. $$ To do this we will look for a root of the function $$ f(x) = x{{\rm e}^{- \left( \sin \left( x/2 \right)  \right) ^{2}}}-3/2.$$
which has the derivative $$f'(x) = {{\rm e}^{- \left( \sin \left( x/2 \right)  \right) ^{2}}}-x\sin
 \left( x/2 \right) \cos \left( x/2 \right) {{\rm e}^{- \left( \sin
 \left( x/2 \right)  \right) ^{2}}}.
$$ The values of $f(x)$ and $f'(x)$ for $x = 2$ have been written in Python for you below so you don't make a mistake copying the formula into you code. 

In [4]:
from math import *

x = 2 # Initial guess

# Define function
f = x * exp(-sin( (1/2) * x)**2) - 3/2 
# Define the derivative of the function (wrt x)
df = exp(-sin((1/2)*x)**2)-x*sin((1/2)*x)*cos((1/2)*x)*exp(-sin((1/2)*x)**2)

print(f"The value of the derivative at x = 2 is f'(2) = {df}")

The value of the derivative at x = 2 is f'(2) = 0.04467938942574401


Notice that the value for the derivative at $x=2$ is very close to zero and is therefore not a good starting point. 

### i) ### 
 There is a root in the interval $[0,10]$. What is the value of this root? Express your answer to 10 decimal places.  

**Note:** As suggested above, the Newton method might not converge for certain initial values, therefore you need to test a few initial starting points until the algorithm converges. 

**Hint:** The below code can be used to print values to many decimal places. To get 10 decimal places of accuracy, you should keep iterating your code until the first 10 decimal places of ${x_n}$ don't change between iterations

In [5]:
n = 1
x = 1/7
f = x*exp(-sin((1/2)*x)**2)-3/2

print(f"n = {n} -- x_n = {x} -- f(x_n) = {f}")
# this prints the integer n, the float x to 10 decimal places
# ... and the float f to 5 decimal places (in exponential format). This 
# ... is best placed inside your loop to see what is happening at each iteration 

n = 1 -- x_n = 0.14285714285714285 -- f(x_n) = -1.357868631509032


**Solution:**

In [7]:
from math import *

x = 5           # note that if you use starting points less than about 2 or higher than about 7 the method will not converge
tol = 1e-12     # User-specified tolerance    
k = 0           # Variable to keep track of the number of iterations


f = x*exp(-sin((1/2)*x)**2)-3/2

while abs(f) > tol and k < 100:
    k = k + 1
    f = x * exp(-sin((1/2)*x)**2)-3/2 
    df = exp(-sin((1/2)*x)**2)-x*sin((1/2)*x)*cos((1/2)*x)*exp(-sin((1/2)*x)**2)
    x = x - f / df
    print(f"n = {k} -- x_n = {x} -- f(x_n) = {f}")
    
print("--- Convergance! ---")
print(f"A root of f(x) is x = {round(x,10)} to 10 decimal places")


n = 1 -- x_n = 4.1599416780402585 -- f(x_n) = 1.9947750322731128
n = 2 -- x_n = 3.8189040617888232 -- f(x_n) = 0.44082866940157994
n = 3 -- x_n = 3.7426244239019977 -- f(x_n) = 0.06883491194650082
n = 4 -- x_n = 3.7390898757335913 -- f(x_n) = 0.0029213575892721444
n = 5 -- x_n = 3.739082543636181 -- f(x_n) = 6.035062285913995e-06
n = 6 -- x_n = 3.7390825436046855 -- f(x_n) = 2.592392966960233e-11
n = 7 -- x_n = 3.7390825436046855 -- f(x_n) = 0.0
--- Convergance! ---
A root of f(x) is x = 3.7390825436 to 10 decimal places


### ii) (Optional bonus question for an extra reward*!) ###

As you may have noticed, Newton's method sometimes doesn't converge unless we are close enough to the solution. One very common method to cirmumvent this issue is to do a few bisection method iterations first, and when you are "close enough" to the solution you can bring it home with Newton iterations. 

Implement a root finding algorithm that:

   (1) uses the bisection method until you are within $|f(c)|<\texttt{tol1}$, then
    
   (2) uses Newton's method until  $|f(x_n)|<\texttt{tol2}$, 
    
where you can choose the values of $\texttt{tol1}$ and $\texttt{tol2}$ as long as $\texttt{tol1}>\texttt{tol2}$.

 \* The reward is the satisfaction of completing the hardest part of the assignment

In [11]:
from math import *

tol1 = 1e-2    # Tolerance for the Bisection search
tol2 = 1e-12   # Tolerance for the Newtons method
k = 0          # Variable to keep track of the number of iterations

# Do bisections until we are have staisfied tol1

a  = 0         # Lower bound
b  = 10        # Upper bound
c  = (a+b)/2   # Bisection 
fc = c * exp(-sin((1/2)*c)**2)-3/2 # f(c)

while abs(fc) > tol1 and k < 20:
    k  = k + 1
    c  = (a+b)/2                                # bisect interval 
    fa = a*exp(-sin((1/2)*a)**2)-3/2  # f(a)
    fb = b*exp(-sin((1/2)*b)**2)-3/2  # f(b)
    fc = c*exp(-sin((1/2)*c)**2)-3/2  # f(c)

    if fa*fc<0:                                 # if f(a) and f(c) have opposite signs then b = c
        b = c

    elif fb*fc<0:                               # if f(a) and f(c) have opposite signs then a = c
        a = c
    print(f"n = {k} -- x_n = {c} -- f(x_n) = {round(fc,10)}")
print("----------------------------------------------------------------------------")    
print(f"Finished bisecting! Now do Newton iterations with starting value x = {round(c,3)}")
print("----------------------------------------------------------------------------")
# use the final value for c to initialize the Newton iterations
x = c
f = x*exp(-sin((1/2)*x)**2)-3/2

while abs(f) > tol and k < 100:
    k = k + 1
    f = x*exp(-sin((1/2)*x)**2)-3/2 
    df = exp(-sin((1/2)*x)**2)-x*sin((1/2)*x)*cos((1/2)*x)*exp(-sin((1/2)*x)**2)
    x = x - f/df
    print(f"n = {k} -- x_n = {x} -- f(x_n) = {f}") 


print("----------------------------------------------------------------------------")    
print(f"Newtons method converged! With the root x = {round(x,10)}")
print("----------------------------------------------------------------------------")

n = 1 -- x_n = 5.0 -- f(x_n) = 1.9947750323
n = 2 -- x_n = 2.5 -- f(x_n) = -0.4841568837
n = 3 -- x_n = 3.75 -- f(x_n) = 0.0090437338
----------------------------------------------------------------------------
Finished bisecting! Now do Newton iterations with starting value x = 3.75
----------------------------------------------------------------------------
n = 4 -- x_n = 3.7391518653696187 -- f(x_n) = 0.009043733757714723
n = 5 -- x_n = 3.7390825464198874 -- f(x_n) = 5.706069388033974e-05
n = 6 -- x_n = 3.7390825436046855 -- f(x_n) = 2.3171775609398537e-09
n = 7 -- x_n = 3.7390825436046855 -- f(x_n) = 0.0
----------------------------------------------------------------------------
Newtons method converged! With the root x = 3.7390825436
----------------------------------------------------------------------------
